tools/defrag_mem_test.py

```diff
@@ -0,0 +1,167 @@
+    print("--------------------------------------------------------------")
+
+
+async def stats_check(connection, condition):
+    try:
+        defrag_task_invocation_total = 0;
+        runs=0
+        while condition.dont_stop():
+            await asyncio.sleep(0.3)
+            done, d = await handle_defrag_stats(connection, defrag_task_invocation_total)
+            if done:
+                print("defrag task successfully found memory locations to reallocate")
+                condition.stop()
+            else:
+                if d is not None:
+                    defrag_task_invocation_total = d
+            runs += 1
+            if runs % 3 == 0:
+                await memory_stats(connection)
+        for i in range(5):
+            done, d = await handle_defrag_stats(connection, -1)
+            if done:
+                print("defrag task successfully found memory locations to reallocate")
+                return True
+            else:
+                await asyncio.sleep(2)
+        return True
+    except Exception as e:
+        print(f"failed to run monitor task: {e}")
+    return False
+
+
+async def delete_keys(connection, keys):
+    results = await connection.delete(*keys)
+    return results
+
+def generate_keys(pattern: str, count: int, batch_size: int) -> list:
+    for i in range(1, count, batch_size):
+        batch = [f"{pattern}{j}" for j in range(i, batch_size + i, 3)]
+        yield batch
+
+async def mem_cleanup(connection, pattern, num, cond, keys_count):
+    counter=0
+    for keys in generate_keys(pattern=pattern, count=keys_count, batch_size=950):
+        if cond.dont_stop() == False:
+            print(f"task number {num} that deleted keys {pattern} finished")
+            return counter
+        counter += await delete_keys(connection, keys)
+        await asyncio.sleep(0.2)
+    print(f"task number {num} that deleted keys {pattern} finished")
+    return counter
+
+
+async def run_tasks(pool, key_name, value_size, keys_count):
+    keys = [f"{key_name}-{i}" for i in range(4)]
+    stop_cond = TaskCancel()
+    try:
+        connection = aioredis.Redis(connection_pool=pool)
+        for key in keys:
+            print(f"creating key {key} with size {value_size} of count {keys_count}")
+            await connection.execute_command("DEBUG", "POPULATE", keys_count, key, value_size)
+            await asyncio.sleep(2)
+        tasks = []
+        count = 0
+        for key in keys:
+            pattern=f"{key}:"
+            print(f"deleting keys from {pattern}")
+            tasks.append(mem_cleanup(connection=connection, pattern=pattern, num=count, cond=stop_cond, keys_count=int(keys_count)))
+            count += 1
+        monitor_task = asyncio.create_task(stats_check(connection, stop_cond))
+        total = await asyncio.gather(*tasks, return_exceptions=True)
+        print(f"successfully deleted {sum(total)} keys")
+        stop_cond.stop()
+        await monitor_task
```

Comment:
You can just use `monitor_task.cancel()` https://docs.python.org/3/library/asyncio-task.html#asyncio.Task.cancel

Reply:
I was thinking about this, the issue in my case is that I want the monitor task to notify to the other tasks as well that there is no need to continue and deleting keys. So this is not really covered with the cancel, in which case, I will rather have my own control for stopping the tasks. I also didn't like the exception base (but maybe this is because I was not using exceptions enough?). Anyway I don't think that in this case (op out as soon as possible) cancel is enough.